pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -117,90 +121,194 @@ func (n *NetworkPolicyController) deleteCNP(old interface{}) {
+	processRules := func(cnpRules []secv1alpha1.Rule, direction controlplane.Direction) {
+		for idx, cnpRule := range cnpRules {
+			services, namedPortExists := toAntreaServicesForCRD(cnpRule.Ports)
+			clusterPeers, perNSPeers := splitPeersByScope(cnpRule, direction)
+			addRule := func(peer *controlplane.NetworkPolicyPeer, dir controlplane.Direction, ruleAppliedTos []string) {
+				rule := controlplane.NetworkPolicyRule{
+					Direction:       dir,
+					Services:        services,
+					Name:            cnpRule.Name,
+					Action:          cnpRule.Action,
+					Priority:        int32(idx),
+					EnableLogging:   cnpRule.EnableLogging,
+					AppliedToGroups: ruleAppliedTos,
+				}
+				if dir == controlplane.DirectionIn {
+					rule.From = *peer
+				} else if dir == controlplane.DirectionOut {
+					rule.To = *peer
+				}
+				rules = append(rules, rule)
 			}
-		}
-		rules = append(rules, controlplane.NetworkPolicyRule{
-			Direction:       controlplane.DirectionIn,
-			From:            *n.toAntreaPeerForCRD(ingressRule.From, cnp, controlplane.DirectionIn, namedPortExists),
-			Services:        services,
-			Name:            ingressRule.Name,
-			Action:          ingressRule.Action,
-			Priority:        int32(idx),
-			EnableLogging:   ingressRule.EnableLogging,
-			AppliedToGroups: appliedToGroupNamesForRule,
-		})
-	}
-	// Compute NetworkPolicyRule for Egress Rule.
-	for idx, egressRule := range cnp.Spec.Egress {
-		// Set default action to ALLOW to allow traffic.
-		services, namedPortExists := toAntreaServicesForCRD(egressRule.Ports)
-		var appliedToGroupNamesForRule []string
-		// Create AppliedToGroup for each AppliedTo present in the ingress rule.
-		for _, at := range egressRule.AppliedTo {
-			var atGroup string
-			if at.Group != "" {
-				atGroup = n.processAppliedToGroupForCG(at.Group)
-			} else {
-				atGroup = n.createAppliedToGroup("", at.PodSelector, at.NamespaceSelector, at.ExternalEntitySelector)
+			if len(clusterPeers) > 0 {
+				ruleAppliedTos := cnpRule.AppliedTo
+				// For ACNPs that have per-namespace rules, cluster-level rules will be created with appliedTo
+				// set as the spec appliedTo for each rule.
+				if appliedToPerRule && len(cnp.Spec.AppliedTo) > 0 {
+					ruleAppliedTos = cnp.Spec.AppliedTo
+				}
+				ruleATGNames := n.processClusterAppliedTo(ruleAppliedTos, atgNamesSet)
+				klog.V(4).Infof("Adding a new cluster-level rule with appliedTos %v for %s", ruleATGNames, cnp.Name)
+				addRule(n.toAntreaPeerForCRD(clusterPeers, cnp, direction, namedPortExists), direction, ruleATGNames)
 			}
-			if atGroup != "" {
-				appliedToGroupNamesForRule = append(appliedToGroupNamesForRule, atGroup)
-				appliedToGroupNamesSet.Insert(atGroup)
+			if len(perNSPeers) > 0 {
+				ruleAppliedTos := cnp.Spec.AppliedTo
+				if len(cnpRule.AppliedTo) > 0 {
+					ruleAppliedTos = cnpRule.AppliedTo
+				}
+				for _, at := range ruleAppliedTos {
+					affectedNS, selectors := n.getAffectedNamespacesForAppliedTo(at)
```

Comment:
nit: in the "common" case of `cnp.Spec.AppliedTo`, maybe we could avoid recomputing this once per rule?

Reply:
Thanks for the suggestion. Done.